### ISSUE TITLE
pre of continuous time variable

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#695]
**Modified by dietmarw on 2 Feb 2012 14:55 UTC**
The specification should be more precise about the semantics of pre(x) in instantaneous equations (reinit or equality inside a when clause). For the following model M1, I noticed that replacing x by pre(x) in the second argument of reinit, or in the right hand side of "d = x" gives exactly the same results at least on one Modelica tool. Does it mean that x and pre(x) are equal inside when clauses for continuous time variables ??

```
model M1
  Real x;
  discrete Real d;
initial equation 
  x = 0;
  d = 0;
equation 
  der(x) = 1;
  when sample(1, 1) then
    reinit(x, x + 1);
    d = x;
  end when;
end M1;
```

----

**Reported by eldjillali&#x2e;talbi on 2 Feb 2012 14:43 UTC**
The specification should be more precise about the semantics of pre(x) in instantaneous equations (reinit or equality inside a when clause). For the following model M1, I noticed that replacing x by pre(x) in the second argument of reinit, or in the right hand side of "d = x" gives exactly the same results at least on one Modelica tool. Does it mean that x and pre(x) are equal inside when clauses for continuous time variables ??

model M1
  Real x;
  discrete Real d;
initial equation 
  x = 0;
  d = 0;
equation 
  der(x) = 1;
  when sample(1, 1) then
    reinit(x, x + 1);
    d = x;
  end when;
end M1;

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/695

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 2 Feb 2012 14:55 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Jun 2013 14:38 UTC**
Will be clarified with the new reinit-definition in #578 - since for a state without reinit pre(x)==x will always be true.

Basically - variant 1:

Event iteration 1 at time=1:
x=pre(x)=1
d=1
Update x to 2
\--
Event iteration 2 at time=1
pre(x)=1  (but not used)
x=2
When is not triggered (no edge) so no reinit and d keeps value d=1
\---
Event iteration 3 at time=1
pre(x)=2 (but not used)
x=2
When is not triggered (no edge) so no reinit and d keeps value d=1
Done since x==pre(x)

Basically - variant 2:

Event iteration 1 at time=1:
x=pre(x)=1
d=1
Update x to 2
pre(x)=x
\--
Event iteration 2 at time=1
pre(x)=2  (but not used)
x=2
When is not triggered (no edge) so no reinit and d keeps value d=1
Done since x==pre(x)

**Comment by modelica-trac-importer** [comment]
**Changelog removed by hansolsson on 4 Jun 2013 14:38 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 24 Jun 2013 07:57 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 24 Jun 2013 12:47 UTC**
Note, x and pre(x) is the same for a state x (which is the case in your example, since x is a state due to "der(x) = 1") with one potential exception: reinit(x,..) changes the state and therefore there might be a difference.

However, the new definition of reinit in Modelica 3.2 rev. 2, treats reinit(x, expr) such that a new variable, say "new(x) = expr" is assigned by "expr". At the end of the model evaluation the equations "x = new(x)" and "pre(x) = x" are then present.

As a result, using x or pre(x) on a state x will always give exactly the same result (and this should be the case in all tools).
